documenation update (#254)
AmigaPorts/ACE: docs/programming/using_bobs.md

viewProcessManagers(s_pView);
copProcessBlocks();
vPortWaitForEnd(s_pVp);
```

## What are BOBs?

BOBs are 2D graphic elements that can be efficiently moved around the screen using the Amiga's hardware blitter. They offer several advantages:

- Background preservation for clean undrawing
- Transparency support via masks
- Animation capabilities by switching frames
- Efficient memory usage and blitter operations
- Support for double-buffering for smooth animation

## Basic BOB Workflow

The typical workflow for using BOBs is:

1. **Initialization** (in gamestate create):
```c
// Create manager with your buffers
bobManagerCreate(pFrontBuffer, pBackBuffer, uwAvailHeight);

// Initialize each BOB
bobInit(&sBob1, uwWidth, uwHeight, 1, pFrameData, pMaskData, uwX, uwY);
bobInit(&sBob2, uwWidth, uwHeight, 1, pFrameData, pMaskData, uwX, uwY);
// ... initialize more BOBs as needed

// Allocate memory for background saving (must be called after all BOBs are initialized)
bobReallocateBuffers();
```

2. **Game Loop** (in gamestate loop):
```c
// Begin the BOB drawing cycle - undraws previous BOBs
bobBegin();

// Update BOB positions
sBob1.sPos.uwX = newX;
sBob1.sPos.uwY = newY;

> // Add BOBs to drawing queue

Reviewer: ```suggestion
// Add BOBs to drawing queue and try to draw the next pending BOB - note that this won't do the draw if blitter is currently busy
```

bobPush(&sBob1);
// ... perform other calculations
bobPush(&sBob2);

> // Process some BOBs (save backgrounds) while doing other calculations

Reviewer: ```suggestion
// Try to draw next bob while doing other calculations, if the blitter isn't busy
```

bobProcessNext();

> // Signal no more BOBs will be added this frame

Reviewer: ```suggestion
// Signal no more BOBs will be added this frame - when using undraw buffers, it will signalize the manager that no more background saves are needed and it can start drawing BOBs on the bitmap with bobProcessNext()
```

bobPushingDone();

// Process remaining BOBs (draw them)

> while(bobProcessNext()) {}

Reviewer: ```suggestion
while(bobProcessNext()) {
  // do some extra work
}
```

// Or simply call bobEnd() to finish all processing
bobEnd();
```

3. **Cleanup** (in gamestate destroy):
```c
bobManagerDestroy();
```

## Important BOB Functions

### Initialization and Setup
- `bobManagerCreate()`: Initializes the BOB system with front/back buffers
- `bobInit()`: Sets up a BOB with its dimensions, position, and frame data
- `bobReallocateBuffers()`: Allocates memory for background saving

### Animation and Dimensions
- `bobSetFrame()`: Changes a BOB's animation frame
- `bobSetWidth()`: Modifies a BOB's width (be careful with memory allocation)
- `bobSetHeight()`: Modifies a BOB's height (be careful with memory allocation)

> - `bobCalcFrameAddress()`: Helps calculate the memory address of a specific animation frame

Reviewer: ```suggestion
- `bobCalcFrameAddress()`: Helps calculate the memory address of a specific animation frame - you can do those beforehand and store the results in the array to get rid of slow multiplications in your game loop
```

### Drawing Cycle
- `bobBegin()`: Starts a new drawing cycle, undraws previous BOBs
- `bobPush()`: Adds a BOB to the drawing queue
- `bobProcessNext()`: Processes the next BOB operation (background save or drawing)
- `bobPushingDone()`: Signals no more BOBs will be added this frame
- `bobEnd()`: Completes all drawing operations

> - `bobDiscardUndraw()`: Skips undrawing (useful for first frame)

Reviewer: ```suggestion
- `bobDiscardUndraw()`: Skips undrawing (useful if you're e.g. redrawing the whole background and don't want the BOB restore to contaminate it with previous one)
```

## Tips and Best Practices

> 1. **Memory Management**:
>    - For BOBs with background preservation (isUndrawRequired=1), the initial width/height should be set to the maximum anticipated size.
>    - Changing width/height to larger than initial size can cause memory corruption.

Reviewer: I'd move it somewhere near describing `bobReallocateBuffers()` and remove this section

```suggestion
> [!NOTE]
> When using the undraw buffers mode, for BOBs with background preservation (`isUndrawRequired` set to `1`) and varying width/height, the initial width/height should be set to the maximum anticipated size. Failing to do so will result in memory corruption.
```

2. **Performance Optimization**:
   - Interleave blitter operations with CPU work by calling `bobProcessNext()` periodically.

>    - Don't use the blitter for other operations between `bobBegin()` and `bobEnd()`.

Reviewer: This is not a performance optimization, it is a requirement - otherwise the whole thing will explode

>    - For non-scrolling games, consider disabling Y-wrapping for better performance.

Reviewer: ```suggestion
   - For games using simpleBuffer, you can safely disable Y-wrapping for better performance - it is only needed for scrollBuffer and tileBuffer-based games.
```

3. **Animation**:

>    - Store animation frames sequentially in memory for easy frame changes.
>    - Use `bobCalcFrameAddress()` to find the correct memory address for each frame.

Reviewer: This needs a move somewhere where the `bobSetFrame()` is used

```suggestion
   - When storing multiple frames in a single bitmap, they need to be put in a single column, one beneath the other.
   - Use `bobCalcFrameAddress()` to calculate the correct memory address for each frame for use with `bobSetFrame()`.
```

> 4. **Double Buffering**:
>    - The BOB system is designed to work with double-buffered displays.
>    - You can use a single buffer by passing the same pointer for front and back buffers.

Reviewer: Move somewhere earlier near the bobManagerCreate() usage or describing a need for a double buffer, remove this section

```suggestion
4. **Double Buffering**:
> [!NOTE]
> Although the BOB system is designed to work with double-buffered displays, you can use a single buffer by passing the same pointer for front and back buffers. The results will probably be unsatisfactory, though.
```

5. **First Frame Handling**:

>    - Call `bobDiscardUndraw()` on the first frame to avoid undrawing BOBs that weren't previously drawn.

Reviewer: This is false

## The BOB Structure

```c
typedef struct tBob {
    UBYTE *pFrameData;      // Current frame bitmap data
    UBYTE *pMaskData;       // Transparency mask data
    tUwCoordYX pOldPositions[2]; // Previous positions for undrawing
    tUwCoordYX sPos;        // Current position (x,y)
    UWORD uwWidth;          // BOB width
    UWORD uwHeight;         // BOB height
    UBYTE isUndrawRequired; // If 1, background is preserved and restored
    // Private fields follow...
} tBob;

> ```

Reviewer: I wouldn't show the bob struct here since it'll change across commits and noone will update it here.

The sPos change is true, but I'll move it to same chapter as bobSetFrame() and call it "Animating BOBs" or something
